File: port/linux/package/pikascript/pikascript-lib/pika_posixfs/pika_posixfs.c
#include "PikaObj.h"
/* simulate the posixfs on linux */
#include <fcntl.h>
#include <unistd.h>

#define sys_open open
#define sys_close close
#define sys_read read
#define sys_write write
#define sys_lseek lseek

int open(const char* __path, int __oflag, ...);
int close(int __fd);
ssize_t read(int __fd, void* __buf, size_t __nbytes);
ssize_t write(int __fd, const void* __buf, size_t __nbytes);

typedef struct _INNER_FILE {
    int fd;
} _INNER_FILE;

/* private prototypes */
int __fmodeflags(const char* mode);
FILE* __fdopen(int fd, const char* mode);

/* public prototypes */
FILE* posixfs_fopen(const char* filename, const char* modes);
size_t posixfs_fwrite(const void* ptr, size_t size, size_t n, FILE* stream);
size_t posixfs_fread(void* ptr, size_t size, size_t n, FILE* stream);
int posixfs_fclose(FILE* stream);
int posixfs_fseek(FILE* stream, long offset, int whence);
long posixfs_ftell(FILE* stream);

#ifndef PIKA_POSIXFS_ENABLE
#define PIKA_POSIXFS_ENABLE 0
#endif

#if PIKA_POSIXFS_ENABLE
FILE* __platform_fopen(const char* filename, const char* modes) {
    return posixfs_fopen(filename, modes);
}
int __platform_fclose(FILE* stream) {
    return posixfs_fclose(stream);
}
size_t __platform_fwrite(const void* ptr, size_t size, size_t n, FILE* stream) {
    return posixfs_fwrite(ptr, size, n, stream);
}
size_t __platform_fread(void* ptr, size_t size, size_t n, FILE* stream) {
    return posixfs_fread(ptr, size, n, stream);
}
int __platform_fseek(FILE* stream, long offset, int whence) {
    return posixfs_fseek(stream, offset, whence);
}
long __platform_ftell(FILE* stream) {
    return posixfs_ftell(stream);
}
#endif

FILE* posixfs_fopen(const char* filename, const char* modes) {
    FILE* f;
    int fd;
    int flags;

    /* Check for valid initial mode character */
    if (!strchr("rwa", *modes)) {
        return 0;
    }

    /* Compute the flags to pass to open() */
    flags = __fmodeflags(modes);

    fd = sys_open(filename, flags, 0666);
    if (fd < 0) {
        return 0;
    }

    f = __fdopen(fd, modes);
    if (f) {
        return f;
    }

    sys_close(fd);
    return 0;
}

size_t posixfs_fwrite(const void* ptr, size_t size, size_t n, FILE* stream) {
    _INNER_FILE* _f = (_INNER_FILE*)stream;
    return sys_write(_f->fd, ptr, n * size);
}

size_t posixfs_fread(void* ptr, size_t size, size_t n, FILE* stream) {
    _INNER_FILE* _f = (_INNER_FILE*)stream;
    return sys_read(_f->fd, ptr, n * size);
}

int posixfs_fclose(FILE* stream) {
    _INNER_FILE* _f = (_INNER_FILE*)stream;
    int fd = _f->fd;
    free(_f);
    return sys_close(fd);
}

int posixfs_fseek(FILE* stream, long offset, int whence) {
    _INNER_FILE* _f = (_INNER_FILE*)stream;
    return sys_lseek(_f->fd, offset, whence);
}

long posixfs_ftell(FILE* stream) {
    _INNER_FILE* _f = (_INNER_FILE*)stream;
    return sys_lseek(_f->fd, 0, 1);
}

int __fmodeflags(const char* mode) {
    int flags;
    if (strchr(mode, '+'))
        flags = O_RDWR;
    else if (*mode == 'r')
        flags = O_RDONLY;
    else
        flags = O_WRONLY;
#if 0
    if (strchr(mode, 'x'))
        flags |= O_EXCL;
    if (strchr(mode, 'e'))
        flags |= O_CLOEXEC;
#endif
    if (*mode != 'r')
        flags |= O_CREAT;
    if (*mode == 'w')
        flags |= O_TRUNC;
    if (*mode == 'a')
        flags |= O_APPEND;
    return flags;
}

FILE* __fdopen(int fd, const char* mode) {
    _INNER_FILE* _f;
    /* Allocate FILE+buffer or fail */
    if (!(_f = malloc(sizeof(*_f))))
        return 0;

    /* Zero-fill only the struct, not the buffer */
    memset(_f, 0, sizeof(*_f));
    _f->fd = fd;
    FILE* f = (FILE*)_f;
    return f;
}

File: packages/pika_posixfs.py
"""POSIX file wrapper using the PikaPython os capability."""
from os import O_CREAT, O_RDONLY, O_RDWR, O_TRUNC, O_WRONLY
from os import SEEK_CUR, SEEK_END, SEEK_SET
from os import close, lseek, open, read, write


def _flags(mode):
    plus = mode.find('+') >= 0
    if mode[0] == 'r':
        flags = O_RDWR if plus else O_RDONLY
    else:
        flags = O_RDWR if plus else O_WRONLY
        flags = flags | O_CREAT
        if mode[0] == 'w':
            flags = flags | O_TRUNC
    return flags


class File:
    def __init__(self, fd):
        self.fd = fd
        self.closed = False

    def read(self, size=-1):
        if size < 0:
            size = 1048576
        return read(self.fd, size)

    def write(self, data):
        return write(self.fd, data)

    def seek(self, offset, whence=SEEK_SET):
        return lseek(self.fd, offset, whence)

    def tell(self):
        return lseek(self.fd, 0, SEEK_CUR)

    def close(self):
        if not self.closed:
            close(self.fd)
            self.closed = True


def fopen(filename, mode='r'):
    return File(open(filename, _flags(mode)))


def fclose(stream):
    stream.close()
    return 0


def fread(size, count, stream):
    return stream.read(size * count)


def fwrite(data, size, count, stream):
    written = stream.write(data[:size * count])
    if size == 0:
        return 0
    return written // size


def fseek(stream, offset, whence=SEEK_SET):
    stream.seek(offset, whence)
    return 0


def ftell(stream):
    return stream.tell()
